File: backend/app/schemas/simulation.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

ALLOWED_SEVERITIES = frozenset({"info", "low", "medium", "high", "critical"})
MAX_CUSTOM_STEPS = 20


class MitreStepInfo(BaseModel):
    technique_id: str
    tactic: str
    name: str


class ScenarioStepResponse(BaseModel):
    order: int
    event_type: str
    offset_seconds: int
    description: str | None = None
    mitre: MitreStepInfo | None = None


class ScenarioResponse(BaseModel):
    id: str
    name: str
    summary: str
    difficulty: str
    event_count: int
    duration_seconds: int
    steps: list[ScenarioStepResponse]
    expected_alerts: list[str]
    expected_outcomes: list[str]


class ScenariosListResponse(BaseModel):
    scenarios: list[ScenarioResponse]
    enabled: bool


class EventTypeOption(BaseModel):
    event_type: str
    category: str


class EventTypesResponse(BaseModel):
    event_types: list[EventTypeOption]


class CustomSimulationStep(BaseModel):
    event_type: str
    offset_seconds: int = Field(ge=0, le=3600)
    severity: str | None = None
    description: str | None = None

    @field_validator("severity")
    @classmethod
    def validate_severity(cls, v: str | None) -> str | None:
        if v is not None and v not in ALLOWED_SEVERITIES:
            raise ValueError(f"severity must be one of {sorted(ALLOWED_SEVERITIES)}")
        return v


class CustomSimulationRequest(BaseModel):
    host_id: UUID
    name: str = Field(min_length=1, max_length=255)
    steps: list[CustomSimulationStep] = Field(min_length=1, max_length=MAX_CUSTOM_STEPS)


class SimulationStepInput(BaseModel):
    event_type: str
    offset_seconds: int
    severity: str | None = None
    description: str | None = None


class SimulationRunResponse(BaseModel):
    message: str
    events: int
    run_id: UUID
    host_id: UUID
    scenario: str
    name: str
    event_ids: list[UUID] = Field(default_factory=list)
    alert_ids: list[UUID] = Field(default_factory=list)
    timeline_ids: list[UUID] = Field(default_factory=list)
    offense_ids: list[UUID] = Field(default_factory=list)


class SimulationRunSummary(BaseModel):
    id: UUID
    scenario_id: str
    name: str
    host_id: UUID
    host_name: str | None = None
    event_count: int
    alert_count: int
    offense_count: int
    timeline_count: int
    created_at: datetime
    run_by: str | None = None

    model_config = {"from_attributes": True}


class SimulationRunListResponse(BaseModel):
    items: list[SimulationRunSummary]
    total: int


class SimulationRunDetail(SimulationRunSummary):
    event_ids: list[UUID] = Field(default_factory=list)
    alert_ids: list[UUID] = Field(default_factory=list)
    timeline_ids: list[UUID] = Field(default_factory=list)
    offense_ids: list[UUID] = Field(default_factory=list)

File: frontend/lib/types/simulation.ts
export interface MitreStepInfo {
  technique_id: string;
  tactic: string;
  name: string;
}

export interface ScenarioStep {
  order: number;
  event_type: string;
  offset_seconds: number;
  description: string | null;
  mitre: MitreStepInfo | null;
}

export interface Scenario {
  id: string;
  name: string;
  summary: string;
  difficulty: "beginner" | "intermediate" | "advanced" | string;
  event_count: number;
  duration_seconds: number;
  steps: ScenarioStep[];
  expected_alerts: string[];
  expected_outcomes: string[];
}

export interface ScenariosResponse {
  scenarios: Scenario[];
  enabled: boolean;
}

export interface EventTypeOption {
  event_type: string;
  category: string;
}

export interface EventTypesResponse {
  event_types: EventTypeOption[];
}

export interface CustomSimulationStep {
  event_type: string;
  offset_seconds: number;
  severity?: string | null;
  description?: string | null;
}

export interface CustomSimulationRequest {
  host_id: string;
  name: string;
  steps: CustomSimulationStep[];
}

export interface SimulationRunResult {
  message: string;
  events: number;
  run_id: string;
  host_id: string;
  scenario: string;
  name: string;
  event_ids: string[];
  alert_ids: string[];
  timeline_ids: string[];
  offense_ids: string[];
}

export interface SimulationRunSummary {
  id: string;
  scenario_id: string;
  name: string;
  host_id: string;
  host_name: string | null;
  event_count: number;
  alert_count: number;
  offense_count: number;
  timeline_count: number;
  created_at: string;
  run_by: string | null;
}

export interface SimulationRunListResponse {
  items: SimulationRunSummary[];
  total: number;
}

export interface SimulationRunDetail extends SimulationRunSummary {
  event_ids: string[];
  alert_ids: string[];
  timeline_ids: string[];
  offense_ids: string[];
}

export type AttackLabTab = "presets" | "custom" | "history";

export interface CustomStepDraft {
  id: string;
  event_type: string;
  offset_seconds: number;
  severity: string;
}
